src/core/global.c

```diff
@@ -570,27 +563,46 @@ int nn_socket (int domain, int protocol)
 int nn_close (int s)
 {
     int rc;
+    struct nn_sock *sock;
 
-    NN_BASIC_CHECKS;
-
-    /*  TODO: nn_sock_term can take a long time to accomplish. It should not be
-        performed under global critical section. */
     nn_glock_lock ();
+    rc = nn_global_hold_socket_locked (&sock, s);
+    if (nn_slow (rc < 0)) {
+        nn_glock_unlock ();
+        errno = -rc;
+        return -1;
+    }
 
-    /*  Deallocate the socket object. */
-    rc = nn_sock_term (self.socks [s]);
+    /*  Start the shutdown process on the socket.  This will cause
+        all other socket users, as well as endpoints, to begin cleaning up.
+        This is done with the glock held to ensure that two instances
+        of nn_close can't access the same socket. */
+
+    nn_sock_stop (sock);
+    nn_glock_unlock ();
+
+    /*  We have to drop both the hold we just acquired, as well as
+        the original hold, in order for nn_sock_term to complete. */
+    nn_sock_rele (sock);
+    nn_sock_rele (sock);
```

Comment:
Lines 586 and 587 are duplicated?

Also, what does "rele" mean? Can this be expanded to its full meaning?